motionplan/planManager.go

```diff
@@ -0,0 +1,428 @@
+	if err != nil {
+		return nil, err
+	}
+	opt.AddConstraint(defaultCollisionConstraintName, collisionConstraint)
+
+	// error handling around extracting motion_profile information from map[string]interface{}
+	var motionProfile string
+	profile, ok := planningOpts["motion_profile"]
+	if ok {
+		motionProfile, ok = profile.(string)
+		if !ok {
+			return nil, errors.New("could not interpret motion_profile field as string")
+		}
+	}
+
+	// convert map to json, then to a struct, overwriting present defaults
+	jsonString, err := json.Marshal(planningOpts)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(jsonString, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	var planAlg string
+	alg, ok := planningOpts["planning_alg"]
+	if ok {
+		planAlg, ok = alg.(string)
+		if !ok {
+			return nil, errors.New("could not interpret planning_alg field as string")
+		}
+		switch planAlg {
+		// TODO(pl): make these consts
+		case "cbirrt":
+			opt.PlannerConstructor = newCBiRRTMotionPlanner
+		case "rrtstar":
+			// no motion profiles for RRT*
+			opt.PlannerConstructor = newRRTStarConnectMotionPlanner
+			// TODO(pl): more logic for RRT*?
+			return opt, nil
+		default:
+			// use default, already set
+		}
+	}
+
+	switch motionProfile {
+	case LinearMotionProfile:
+		// Linear constraints
+		linTol, ok := planningOpts["line_tolerance"].(float64)
+		if !ok {
+			// Default
+			linTol = defaultLinearDeviation
+		}
+		orientTol, ok := planningOpts["orient_tolerance"].(float64)
+		if !ok {
+			// Default
+			orientTol = defaultLinearDeviation
+		}
+		constraint, pathDist := NewAbsoluteLinearInterpolatingConstraint(from, to, linTol, orientTol)
+		opt.AddConstraint(defaultLinearConstraintName, constraint)
+		opt.pathDist = pathDist
+	case PseudolinearMotionProfile:
+		tolerance, ok := planningOpts["tolerance"].(float64)
+		if !ok {
+			// Default
+			tolerance = defaultPseudolinearTolerance
+		}
+		constraint, pathDist := NewProportionalLinearInterpolatingConstraint(from, to, tolerance)
+		opt.AddConstraint(defaultPseudolinearConstraintName, constraint)
+		opt.pathDist = pathDist
+	case OrientationMotionProfile:
+		tolerance, ok := planningOpts["tolerance"].(float64)
+		if !ok {
+			// Default
+			tolerance = defaultOrientationDeviation
+		}
+		constraint, pathDist := NewSlerpOrientationConstraint(from, to, tolerance)
+		opt.AddConstraint(defaultOrientationConstraintName, constraint)
+		opt.pathDist = pathDist
+	case PositionOnlyMotionProfile:
+		opt.SetMetric(NewPositionOnlyMetric())
+	case FreeMotionProfile:
+		// No restrictions on motion
+		fallthrough
+	default:
+		if planAlg == "" {
+			// set up deep copy for fallback
+			try1 := deepAtomicCopyMap(planningOpts)
+			// No need to generate tons more IK solutions when the first alg will do it
+			opt.MaxSolutions = defaultFallbackIK
+
+			// extra timeout to account for IK
+			try1["timeout"] = defaultFallbackTimeout
```

Comment:
[q] can you explain what this is doing?

Reply:
That comment isn't correct.

But, this is the timeout that says "run RRT* for X amount of time", before cancelling and handing the map off to cbirrt.

Reply:
cool much clearer now